sdks/java/io/rabbitmq/src/main/java/org/apache/beam/sdk/io/rabbitmq/RabbitMqMessage.java

```diff
@@ -17,31 +17,208 @@
  */
 package org.apache.beam.sdk.io.rabbitmq;
 
+import com.google.auto.value.AutoValue;
 import com.rabbitmq.client.AMQP;
 import com.rabbitmq.client.AMQP.BasicProperties;
 import com.rabbitmq.client.Envelope;
 import com.rabbitmq.client.GetResponse;
 import com.rabbitmq.client.LongString;
 import java.io.Serializable;
-import java.util.Arrays;
+import java.nio.charset.StandardCharsets;
+import java.time.temporal.ChronoUnit;
+import java.util.Collections;
 import java.util.Date;
 import java.util.HashMap;
 import java.util.Map;
-import java.util.Objects;
+import java.util.Optional;
+import java.util.UUID;
 import javax.annotation.Nullable;
+import org.apache.beam.sdk.coders.DefaultCoder;
+import org.apache.beam.sdk.coders.SerializableCoder;
 
 /**
- * It contains the message payload, and additional metadata like routing key or attributes. The main
- * reason of this class is that AMQP.BasicProperties doesn't provide a serializable public API.
+ * Serializable wrapper around rabbitmq's {@link GetResponse}.
+ *
+ * <p>The main purpose of this class is to provide a serializable public API for
+ * AMQP.BasicProperties, which lacks one of its own. Secondarily, introducing the builder pattern
+ * with AutoValue makes this a bit easier to work with.
  */
-public class RabbitMqMessage implements Serializable {
+@AutoValue
+@DefaultCoder(SerializableCoder.class)
+public abstract class RabbitMqMessage implements Serializable {
+  public static final int DELIVERY_MODE_NON_DURABLE = 1;
+  public static final int DELIVERY_MODE_DURABLE = 2;
+
+  public static final int PRIORITY_LOWEST = 1;
+
+  /**
+   * Defaults to the empty string. When publishing messages to an exchange type other than fanout,
+   * this should be specified.
+   */
+  public abstract String routingKey();
+
+  @SuppressWarnings("mutable")
+  public abstract byte[] body();
+
+  @Nullable
+  public abstract String contentType();
+
+  @Nullable
+  public abstract String contentEncoding();
+
+  public abstract Map<String, Object> headers();
```

Comment:
There is an implicit, breaking change here. Previously, `headers` was not included in the `hashCode` or `equals` implementations, but this feels like it was an oversight to me.